text/0000-edition-based-method-disambiguation.md

```diff
@@ -0,0 +1,224 @@
+  --> src/lib.rs:16:5
+   |
+16 |     fn intersperse(self, separator: Self::Item) -> Intersperse<Self> {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   = note: to avoid the ambiguity, this candidate was selected to prevent breakage in edition < 20XX
+help: disambiguate the associated function for candidate #1
+   |
+23 |     Iterator::intersperse(it, sep)
+   |
+help: disambiguate the associated function for candidate #2
+   |
+23 |     Itertools::intersperse(it, sep)
+   |
+```
+
+These errors are an expected stage of the standard library development
+lifecycle, where methods are experimented within 3rd party crates and then
+moved into the standard library once they've been thorougly tested. A classic
+example of this is the `Itertools` crate for experimenting with extentions to
+the `Iterator` trait. However this problem isn't restricted to extension traits
+of pre-edition standard library traits, and can indeed become a problem whenever
+any two methods have the same name.
+
+You can fix issues like this by manually editing the code to select the
+specific version of the method you wish to use or, in certain common cases, you
+can use cargo fix. cargo fix will make assumptions about how the methods relate
+depending on if you're using cargo fix for an edition upgrade or not.
+
+* **Within same edition** cargo fix will assume that the new method is a drop
+  in replacement of the pre-edition downstream one and will disambiguate by
+  selecting the upstream method defined in `std`.
+* **As part of an edition upgrade** cargo fix will prioritize maintaining the
+  same behavior, and will disambiguate by selecting the pre-edition method that
+  was being used previously.
+
+To run cargo fix within the same edition run:
+
+```
+cargo fix
+```
+
+In the example above this would replace the ambiguous code with
+`Iterator::intersperse(it, sep)`, selecting the new implementation.
+
+To run cargo fix as part of an edition upgrade run:
+
+```
+cargo fix --edition
+```
+
+In the example above this would replace the ambiguous code with
+`Itertools::intersperse(it, sep)`, maintaining the pre-edition behavior.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be implemented by modifying the `rustc_stable` attribute to
+support an additional optional `edition` field.
+
+During method resolution, when we detect an ambiguity we should then check if
+one of the methods in question is a standard libary method with an `edition`
+field. When the edition field exists in the stability attribute and the edition
+field of that method matches the current crate's edition we ignore that method
+and select the pre-edition method that conflicted with it and generate a
+warning. If the edition field in the stability attribute is an earlier edition
+than the crate's edition we continue as normal and emit an error for the
+ambiguity.
```

Comment:
Seems plausible, but as noted elsewhere this reasoning would like only apply to one level of autoderef. An alternative might be:

* If a method with this "dependent" character exists and we are in the "compatibility" mode:
    * Run the method selection probe once without the method in the list.
    * If this gets a "no such method" error, run again WITH the method in the list.

This seems like it would be strictly non-breaking. 

My hunch is that it makes sense -- at this RFC stage -- to leave this unspecified and instead to focus the RFC on enumerated interesting examples and how we would ideally like them to behave. I can come up with various patterns to think about. Then we worry about how to implement it.